src/hw/hw_mngr.cpp

```diff
@@ -357,3 +509,177 @@ void HWManager::setFanSpeedDelayed()
         setFanSpeed();
     });
 }
+
+void HWManager::detectFansDelayed(uint32_t delaySecs)
+{
+    static boost::asio::deadline_timer tmr(io);
+
+    //  if we has expired attempts, then stop fan detection 
+    //  and work in normal state
+    if (++numErrorAttempts > maxErrorAttempts)
+    {
+        log<level::ERR>("Reached maxErrorAttempts",
+                        entry("STATE=%d", int(detectFansState)));
+        detectFansState = FanState::normal;
+        return;
+    }
+
+    // this implicitly cancels the timer
+    tmr.expires_from_now(boost::posix_time::seconds(delaySecs));
+    tmr.async_wait([&](const boost::system::error_code& err) {
+        if (err == boost::asio::error::operation_aborted)
+        {
+            /* we were canceled*/
+            return;
+        }
+        else if (err)
+        {
+            log<level::ERR>("Timer error",
+                            entry("WHAT=%s", err.message().c_str()));
+            return;
+        }
+
+        processDetectState();
+    });
+}
+
+void HWManager::onHostPowerOn()
+{
+    log<level::DEBUG>("Host power is ON");
+    detectFansState = FanState::uninit;
+    numErrorAttempts = 0;
+}
+
+void HWManager::onHostPowerOff()
+{
+    log<level::DEBUG>("Host power is OFF");
+}
+
+void HWManager::runDetectFans()
+{
+    if (detectFansState == FanState::uninit)
+    {
+        detectFansState = FanState::init;
+        processDetectState();
+    }
+}
+
+void HWManager::processDetectState()
+{
+    if (detectFansState == FanState::normal)
+        return;
+ 
+    //  wait until createInventory() complete a work
+    if (!config.desc)
+    {
+        detectFansDelayed(5);
+        return;
+    }
+
+    if (!processSystemFans())
+    {
+        detectFansState = FanState::normal;
+        return;
+    }
+
+    switch (detectFansState)
```

Comment:
I prefer to not spread state machine to entire file - it is hard to read, understand and trace.. Can you keep it in one function?

Reply:
Look again,  spread state machine is implemeted in one function - processDetectState()